packages/node/src/indexer/indexer.manager.ts

```diff
@@ -456,84 +445,171 @@ export class IndexerManager {
     );
   }
 
-  private async indexBlockForRuntimeDs(
-    vm: IndexerSandbox,
-    handlers: SubqlRuntimeHandler[],
+  private async indexBlockData(
     { block, events, extrinsics }: BlockContent,
+    dataSources: SubqlProjectDs[],
+    getVM: (d: SubqlProjectDs) => IndexerSandbox,
   ): Promise<void> {
-    for (const handler of handlers) {
-      switch (handler.kind) {
-        case SubqlHandlerKind.Block:
-          if (SubstrateUtil.filterBlock(block, handler.filter)) {
-            await vm.securedExec(handler.handler, [block]);
-          }
-          break;
-        case SubqlHandlerKind.Call: {
-          const filteredExtrinsics = SubstrateUtil.filterExtrinsics(
-            extrinsics,
-            handler.filter,
-          );
-          for (const e of filteredExtrinsics) {
-            await vm.securedExec(handler.handler, [e]);
-          }
-          break;
-        }
-        case SubqlHandlerKind.Event: {
-          const filteredEvents = SubstrateUtil.filterEvents(
-            events,
-            handler.filter,
-          );
-          for (const e of filteredEvents) {
-            await vm.securedExec(handler.handler, [e]);
-          }
-          break;
-        }
-        default:
+    await this.indexBlockContent(block, dataSources, getVM);
+
+    for (const extrinsic of extrinsics) {
```

Comment:
Initialise phase events should be handled before any extrinsics and ApplyExtrinsic phase events.
Finalise phase events and other remaining events are handled last.